src/payment/strategies/bolt/bolt-app-payment-strategy.ts

```diff
@@ -0,0 +1,106 @@
+
+import { BoltCheckout, BoltTransacion } from './bolt';
+import BoltScriptLoader from './bolt-script-loader';
+
+export default class BoltAppPaymentStrategy implements PaymentStrategy {
+    private _boltClient?: BoltCheckout;
+    private _methodId!: string;
+
+    constructor(
+        private _store: CheckoutStore,
+        private _orderActionCreator: OrderActionCreator,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _boltScriptLoader: BoltScriptLoader
+    ) { }
+
+    async initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        this._methodId = options.methodId;
+
+        const state = this._store.getState();
+        const paymentMethod = state.paymentMethods.getPaymentMethod(this._methodId);
+
+        if (!paymentMethod || !paymentMethod.initializationData.publishableKey) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const publishableKey = paymentMethod.initializationData.publishableKey;
+
+        this._boltClient = await this._boltScriptLoader.load(publishableKey, paymentMethod.config.testMode);
+
+        return Promise.resolve(this._store.getState());
+    }
+
+    deinitialize(): Promise<InternalCheckoutSelectors> {
+        this._boltClient = undefined;
+
+        return Promise.resolve(this._store.getState());
+    }
+
+    async execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const { payment, ...order } = payload;
+
+        if (!payment) {
+            throw new InvalidArgumentError('Unable to submit payment because "payload.payment" argument is not provided.');
+        }
+
+        await this._store.dispatch(this._orderActionCreator.submitOrder(order, options));
+
+        const state = await this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(payment.methodId, options));
+        const paymentMethod = state.paymentMethods.getPaymentMethod(payment.methodId);
+
+        if (!paymentMethod || !paymentMethod.clientToken) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        const orderToken = paymentMethod.clientToken;
+
+        const transaction: BoltTransacion = await new Promise((resolve, reject) => {
+            const onSuccess = (transaction: BoltTransacion,  callback: () => void) => {
+                resolve(transaction);
+                callback();
+            };
+
+            const onClose = () => {
+                reject(new PaymentMethodCancelledError());
+            };
+
+            const callbacks = {
+                success: onSuccess,
+                close: onClose,
+            };
+
+            if (!this._boltClient) {
+                throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
```

Comment:
should you reject the promise in here?
```
reject(new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized))`

Reply:
If I validate this outside the promise, it will mask the boltClient as possibly undefined
<img width="811" alt="Screen Shot 2020-07-29 at 12 02 37 PM" src="https://user-images.githubusercontent.com/35502707/88830340-70239800-d193-11ea-836c-4acb8537aaf1.png">